File: backend/core/blog_slug.py
"""Normalización y validación canónica de slugs públicos del blog."""

import unicodedata

MAX_BLOG_SLUG_LENGTH = 150


def normalize_blog_slug(value: object) -> str | None:
    """Devuelve el slug NFC si tiene la misma forma canónica que genera el frontend."""
    if not isinstance(value, str):
        return None

    slug = unicodedata.normalize("NFC", value)
    if not slug or len(slug) > MAX_BLOG_SLUG_LENGTH:
        return None

    previous_was_alphanumeric_or_mark = False
    has_alphanumeric = False

    for character in slug:
        if character == "-":
            # Un slug generado correctamente no empieza por guion, no contiene
            # guiones consecutivos y tampoco termina en guion.
            if not previous_was_alphanumeric_or_mark:
                return None
            previous_was_alphanumeric_or_mark = False
            continue

        if character.isalnum():
            has_alphanumeric = True
            previous_was_alphanumeric_or_mark = True
            continue

        if (
            unicodedata.category(character).startswith("M")
            and previous_was_alphanumeric_or_mark
        ):
            continue

        return None

    if not has_alphanumeric or not previous_was_alphanumeric_or_mark:
        return None

    return slug

File: frontend/src/utils/blogSlug.ts
/** Longitud máxima de la columna `blog.slug` en MySQL. */
export const MAX_BLOG_SLUG_LENGTH = 150;

/**
 * Normaliza un slug y comprueba la misma forma canónica que produce `slugify`:
 * letras/números Unicode, marcas combinadas válidas y guiones simples interiores.
 */
export const normalizeBlogSlug = (value: unknown): string | null => {
  if (typeof value !== "string") {
    return null;
  }

  const slug = value.normalize("NFC");
  const slugLength = Array.from(slug).length;
  if (slugLength === 0 || slugLength > MAX_BLOG_SLUG_LENGTH) {
    return null;
  }

  let previousWasAlphanumericOrMark = false;
  let hasAlphanumeric = false;

  for (const character of slug) {
    if (character === "-") {
      if (!previousWasAlphanumericOrMark) {
        return null;
      }
      previousWasAlphanumericOrMark = false;
      continue;
    }

    if (/^[\p{L}\p{N}]$/u.test(character)) {
      hasAlphanumeric = true;
      previousWasAlphanumericOrMark = true;
      continue;
    }

    if (/^\p{M}$/u.test(character) && previousWasAlphanumericOrMark) {
      continue;
    }

    return null;
  }

  return hasAlphanumeric && previousWasAlphanumericOrMark ? slug : null;
};

export const isValidBlogSlug = (value: unknown): value is string =>
  normalizeBlogSlug(value) !== null;
